core/src/main/java/org/apache/iceberg/io/datafile/DataFileServiceRegistry.java

```diff
@@ -0,0 +1,231 @@
+      throw new IllegalArgumentException(
+          String.format(
+              "Read builder %s clashes with %s. Both serves %s",
+              readBuilder.getClass(), READ_BUILDERS.get(key), key));
+    }
+
+    READ_BUILDERS.put(new Key(format, outputType, readerType), readBuilder);
+  }
+
+  @SuppressWarnings("CatchBlockLogException")
+  private static void registerSupportedFormats() {
+    Avro.register();
+
+    // Uses dynamic methods to call the `register` for the listed classes
+    for (String s : CLASSES_TO_REGISTER) {
+      try {
+        DynMethods.StaticMethod register =
+            DynMethods.builder("register").impl(s).buildStaticChecked();
+
+        register.invoke();
+
+      } catch (NoSuchMethodException e) {
+        // failing to register readers/writers is normal and does not require a stack trace
+        LOG.info("Unable to register {} for data files: {}", s, e.getMessage());
+      }
+    }
+  }
+
+  static {
+    registerSupportedFormats();
+  }
+
+  private DataFileServiceRegistry() {}
+
+  /**
+   * Provides a reader builder for the given input file which returns objects with a given
+   * returnType.
+   *
+   * @param format of the file to read
+   * @param returnType returned by the reader
+   * @param inputFile to read
+   * @param <F> type of the records which are filtered by {@link DeleteFilter}
+   * @return {@link ReadBuilder} for building the actual reader
+   */
+  public static <F> ReadBuilder<?, F> readBuilder(
+      FileFormat format, String returnType, InputFile inputFile) {
+    return readBuilder(format, returnType, null, inputFile);
+  }
+
+  /**
+   * Provides a reader builder for the given input file which returns objects with a given
+   * returnType.
+   *
+   * @param format of the file to read
+   * @param returnType returned by the reader
+   * @param builderType of the reader builder
+   * @param inputFile to read
+   * @param <F> type of the records which are filtered by {@link DeleteFilter}
+   * @return {@link ReadBuilder} for building the actual reader
+   */
+  public static <F> ReadBuilder<?, F> readBuilder(
+      FileFormat format, String returnType, String builderType, InputFile inputFile) {
+    return (ReadBuilder<?, F>)
+        READ_BUILDERS.get(new Key(format, returnType, builderType)).apply(inputFile);
+  }
+
+  /**
+   * Provides a writer builder for the given output file which writes objects with a given
+   * inputType.
+   *
+   * @param format of the file to read
+   * @param inputType accepted by the writer
+   * @param outputFile to write
+   * @param <E> type for the engine specific schema used by the builder
+   * @return {@link ReadBuilder} for building the actual reader
+   */
+  public static <E> WriteBuilder<?, E> writeBuilder(
+      FileFormat format, String inputType, EncryptedOutputFile outputFile) {
+    return new WriteBuilder<>(
+        (AppenderBuilder<?, E>)
+            APPENDER_BUILDERS.get(new Key(format, inputType, null)).apply(outputFile),
```

Comment:
Callers should not need to fill in default values. Can you add a constructor to avoid the caller needing to know how the key represents missing?

Reply:
Callers don't need to provide it, but removed because of another comment